source_code/main_mcu/src/LOGIC/logic_encryption.c

```diff
@@ -252,4 +270,194 @@ void logic_encryption_ctr_decrypt(uint8_t* data, uint8_t* cred_ctr, uint16_t dat
+*/
+void logic_encryption_sha512_update(uint8_t const *data, size_t len)
+{
+    br_sha512_update(&logic_encryption_sha512_ctx, data, len);
+}
+
+/*! \fn     logic_encryption_sha256_final(uint8_t *hash)
+*   \brief  Finalize the hash and out put the value to the argument
+*   \param  hash  The final hash value (output)
+*/
+void logic_encryption_sha256_final(uint8_t *hash)
+{
+    br_sha256_out(&logic_encryption_sha256_ctx, hash);
+}
+
+/*! \fn     logic_encryption_sha512_final(uint8_t *hash)
+*   \brief  Finalize the hash and output the value to the argument
+*   \param  hash  The final hash value (output)
+*/
+void logic_encryption_sha512_final(uint8_t *hash)
+{
+    br_sha512_out(&logic_encryption_sha512_ctx, hash);
+}
+
+/*! \fn     logic_encryption_sha256_hmac_init(uint8_t const *key, uint32_t klen)
+*   \brief  Initialize the sha256 HMAC object
+*   \param  key     key used for HMAC
+*   \param  len     length of key
+*/
+void logic_encryption_sha256_hmac_init(uint8_t const *key, uint32_t klen)
+{
+    br_hmac_key_init(&logic_encryption_hmac_kc, &br_sha256_vtable, key, klen);
+    br_hmac_init(&logic_encryption_hmac_ctx, &logic_encryption_hmac_kc, 0);
+}
+
+/*! \fn     logic_encryption_sha256_hmac_update(uint8_t const *data, size_t len)
+*   \brief  Update the HMAC256 value with the input data
+*   \param  data    buffer containing the data to compute HMAC256 over
+*   \param  len     length of data buffer
+*/
+void logic_encryption_sha256_hmac_update(uint8_t const *data, size_t len)
+{
+    br_hmac_update(&logic_encryption_hmac_ctx, data, len);
+}
+
+/*! \fn     logic_encryption_sha256_hmac_final(uint8_t * hmac_out)
+*   \brief  Finalize the HMAC and output the value to the argument
+*   \param  hash  The final HMAC value (output)
+*/
+void logic_encryption_sha256_hmac_final(uint8_t * hmac_out)
+{
+    br_hmac_out(&logic_encryption_hmac_ctx, hmac_out);
+}
+
+/*! \fn     logic_encryption_ecc256_init(void)
+*   \brief  Initialize the ECC256 crypto object
+*/
+void logic_encryption_ecc256_init(void)
+{
+    #define SEED_LEN 8
+    uint8_t seed[SEED_LEN];
+
+    logic_encryption_br_ec_algo = &br_ec_p256_m15;
+    logic_encryption_br_ec_algo_id = BR_EC_secp256r1;
+
+    rng_fill_array(seed, SEED_LEN);
+    br_hmac_drbg_init(&logic_encryption_hmac_drbg_ctx, &br_sha256_vtable, seed, SEED_LEN);
+}
+
+/*! \fn     logic_encryption_ecc256_sign(uint8_t const *data, int len, uint8_t * sig)
+*   \brief  Cryptographically sign input data and return the signature in arg.
+*   \param  data    data to sign
+*   \param  len     length of data to sign
+*   \param  sig     output signature
+*/
+void logic_encryption_ecc256_sign(uint8_t const *data, int len, uint8_t * sig)
+{
+    size_t result = br_ecdsa_i15_sign_raw(logic_encryption_br_ec_algo, logic_encryption_sha256_ctx.vtable, data, &logic_encryption_signing_key, sig);
+
+    if (result == 0)
+    {
+        //TODO: Log(TAG_ERR, "error, signing failed!\n");
+        while(1);
```

Comment:
what could cause the function to fail? if it's something malicious then leave the while()

Reply:
Will look into it how we can best handle this without freezing.

Reply:
Should be programmer error if this occurs. Will leave it.